core/src/SYCL/Kokkos_SYCL_Parallel_Range.hpp

```diff
@@ -63,8 +63,9 @@ class Kokkos::Impl::ParallelFor<FunctorType, ExecPolicy,
   ParallelFor()        = delete;
   ParallelFor& operator=(const ParallelFor&) = delete;
 
-  static void sycl_direct_launch(const Policy& policy,
-                                 const FunctorType& functor) {
+  template <typename PolicyType, typename Functor>
+  static void sycl_direct_launch(const PolicyType& policy,
+                                 const Functor& functor) {
```

Comment:
The commit message says " template ParallelFor::sycl_direct_launch to avoid unconditional instantiation". If the function is not templated, it is unconditionally (implicitly) instantiated when the class is instantiated and the compiler complains that we try to launch a kernel that is non-trivially copyable, essentially the issue we are trying to work around with the indirect launch mechanism in the first place.

Reply:
I would prefer
```C++
template <typename AlwaysVoid = void>
static void sycl_direct_launch(...)
```
because I think coexistence of `Policy`, `PolicyType`, `Functor`, and `FunctorType` is confusing.

Reply:
That doesn't quite work since I still need the type of the passed functor to be trivially-copyable. I can omit the template parameter for the policy, though.